pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+
+			let collateral = info.collateral();
+			T::NonFungible::transfer(&collateral.0, &collateral.1, &who)?;
+
+			ClosedLoans::<T>::insert(pool_id, loan_id, ClosedLoan::new(info)?);
+
+			Self::deposit_event(Event::<T>::Closed {
+				pool_id,
+				loan_id,
+				collateral,
+			});
+
+			Ok(())
+		}
+
+		/// Updates the write off policy.
+		///
+		/// The write off policy is used to automatically set a write off minimum value to the
+		/// loan.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn update_write_off_policy(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			mut policy: BoundedVec<WriteOffState<T::Rate>, T::MaxWriteOffGroups>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::ensure_role(pool_id, &who, PoolRole::LoanAdmin)?;
+			Self::ensure_pool_exists(pool_id)?;
+
+			policy.iter_mut().try_for_each(|state| -> DispatchResult {
+				state.penalty = Self::to_rate_per_sec(state.penalty)?;
+				Ok(())
+			})?;
+
+			WriteOffPolicy::<T>::insert(pool_id, policy);
+
+			Self::deposit_event(Event::<T>::WriteOffPoliciesUpdated { pool_id });
+
+			Ok(())
+		}
+
+		/// Updates the porfolio valuation for the given pool
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn update_portfolio_valuation(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			Self::ensure_pool_exists(pool_id)?;
+
+			let value = Self::portfolio_valuation_for_pool(pool_id)?;
+
+			LatestPortfolioValuations::<T>::insert(
+				pool_id,
+				PortfolioValuation::new(value, Self::now()),
+			);
+
+			Self::deposit_event(Event::<T>::PortfolioValuationUpdated {
+				pool_id,
+				value,
+				update_type: PortfolioValuationUpdateType::Exact,
+			});
+
+			Ok(())
+		}
+	}
+
+	/// Utility methods
+	impl<T: Config> Pallet<T> {
+		fn now() -> Moment {
+			T::Time::now().as_secs()
+		}
+
+		fn ensure_role(pool_id: PoolIdOf<T>, who: &T::AccountId, role: PoolRole) -> DispatchResult {
+			T::Permissions::has(
+				PermissionScope::Pool(pool_id),
+				who.clone(),
+				Role::PoolRole(role),
+			)
+			.then_some(())
+			.ok_or(BadOrigin.into())
+		}
+
+		fn ensure_collateral_owner(
+			owner: &T::AccountId,
+			(collection_id, item_id): AssetOf<T>,
+		) -> DispatchResult {
+			T::NonFungible::owner(&collection_id, &item_id)
+				.ok_or(Error::<T>::NFTOwnerNotFound)?
+				.eq(owner)
+				.then_some(())
+				.ok_or(Error::<T>::NotNFTOwner.into())
+		}
+
+		fn ensure_loan_borrower(owner: &T::AccountId, borrower: &T::AccountId) -> DispatchResult {
+			ensure!(owner == borrower, Error::<T>::NotLoanBorrower);
+			Ok(())
+		}
+
+		fn ensure_pool_exists(pool_id: PoolIdOf<T>) -> DispatchResult {
+			ensure!(T::Pool::pool_exists(pool_id), Error::<T>::PoolNotFound);
+			Ok(())
+		}
+
+		fn generate_loan_id() -> T::LoanId {
+			LastLoanId::<T>::mutate(|last_loan_id| {
+				*last_loan_id = T::Hasher::hash(&*last_loan_id.as_ref());
+				*last_loan_id
+			})
+		}
```

Comment:
It could also be an incrementing integer, but I think a hash represents better what an ID means.

Reply:
Currently, this is also used as more of a user facing ID, so an issuer talks about "loan 8". I think there's value in that, so I'd prefer to keep it as an incrementing integer.

Reply:
No worries! I change this

Reply:
Only a question, should this number "loan 8" be per pool, or can it be something global?

Reply:
It's currently per pool and would be great to keep that.

Reply:
Done!